File: Problems/Algorithms/1129. Shortest Path with Alternating Colors/ShortestAlternatingPath.kt
class Solution {
    fun shortestAlternatingPaths(n: Int, redEdges: Array<IntArray>, blueEdges: Array<IntArray>): IntArray {
        val redGraph: MutableMap<Int, MutableList<Int>> = mutableMapOf()
        for (edge in redEdges) {
            if (!redGraph.containsKey(edge[0])) {
                redGraph[edge[0]] = mutableListOf()
            }
            redGraph[edge[0]]!!.add(edge[1])
        }
        
        val blueGraph: MutableMap<Int, MutableList<Int>> = mutableMapOf()
        for (edge in blueEdges) {
            if (!blueGraph.containsKey(edge[0])) {
                blueGraph[edge[0]] = mutableListOf()
            }
            blueGraph[edge[0]]!!.add(edge[1])
        }
        
        val visited: HashSet<String> = hashSetOf()   
        val results = IntArray(n) { -1 }
        
        val queue = ArrayDeque<Pair<Int, String>>()
        queue.add(Pair(0, "null"))
        var steps = 0
        while (!queue.isEmpty()) {
            val length = queue.size
            for (i in 0..length-1) {
                val curr = queue.removeFirst()
                val node = curr.first
                val color = curr.second
                
                val key = node.toString() + "->" + color
                if (visited.contains(key)) {
                    continue
                }
                visited.add(key)
                
                if (results[node] == -1) {
                    results[node] = steps
                }
                
                if (color == "null" || color == "red") {
                    if (blueGraph[node] != null) {
                        for (neighbor in blueGraph[node]!!) {
                            queue.add(Pair(neighbor, "blue"))
                        }
                    } 
                }
                if (color == "null" || color == "blue") {
                    if (redGraph[node] != null) {
                        for (neighbor in redGraph[node]!!) {
                            queue.add(Pair(neighbor, "red"))
                        }
                    }
                }
            }
            
            steps++
        }
        
        return results
    }
}

File: Problems/Algorithms/1129. Shortest Path with Alternating Colors/shortest_alternating_path.py
#!/usr/bin/env python3

from collections import defaultdict


class Solution:
    def shortestAlternatingPaths(self, n: int, redEdges: List[List[int]], blueEdges: List[List[int]]) -> List[int]:
        red_graph = defaultdict(list)
        for edge in redEdges:
            red_graph[edge[0]].append(edge[1])
        
        blue_graph = defaultdict(list)
        for edge in blueEdges:
            blue_graph[edge[0]].append(edge[1])
        
        visited = set()   
        results = [-1] * n

        queue = [(0, "null")]
        steps = 0
        while len(queue) > 0:
            length = len(queue)
            for i in range(length):
                curr = queue.pop(0)
                node, color = curr[0], curr[1]
                
                key = str(node) + "->" + color
                if key in visited:
                    continue
                visited.add(key)
                
                if results[node] == -1:
                    results[node] = steps
                
                if color == "null" or color == "red":
                    if blue_graph[node]:
                        for neighbor in blue_graph[node]:
                            queue.append((neighbor, "blue"))
                if color == "null" or color == "blue":
                    if red_graph[node]:
                        for neighbor in red_graph[node]:
                            queue.append((neighbor, "red"))

            steps += 1
        
        return results
